File: platforms/apple/Sources/BotaAppleSDK/Core/CoreCommand.swift
import BotaDeviceSDKC
import Foundation

struct CoreCapabilities: OptionSet, Equatable, Sendable {
    let rawValue: UInt64

    static let bluetooth = Self(rawValue: UInt64(BOTA_DEVICE_SDK_V1_CAPABILITY_BLE))
    static let timer = Self(rawValue: UInt64(BOTA_DEVICE_SDK_V1_CAPABILITY_TIMER))
    static let persistence = Self(rawValue: UInt64(BOTA_DEVICE_SDK_V1_CAPABILITY_PERSISTENCE))
    static let secureStorage = Self(rawValue: UInt64(BOTA_DEVICE_SDK_V1_CAPABILITY_SECURE_STORAGE))
    static let networkTransfer = Self(rawValue: UInt64(BOTA_DEVICE_SDK_V1_CAPABILITY_NETWORK_TRANSFER))
    static let progress = Self(rawValue: UInt64(BOTA_DEVICE_SDK_V1_CAPABILITY_PROGRESS))
    static let hostMaterial = Self(rawValue: UInt64(BOTA_DEVICE_SDK_V1_CAPABILITY_HOST_MATERIAL))
    static let recordingSink = Self(rawValue: UInt64(BOTA_DEVICE_SDK_V1_CAPABILITY_RECORDING_SINK))
    static let firmwareBlob = Self(rawValue: UInt64(BOTA_DEVICE_SDK_V1_CAPABILITY_FIRMWARE_BLOB))

    init(rawValue: UInt64) {
        self.rawValue = rawValue
    }

    init(names: [String]) throws {
        var value: Self = []
        for name in names {
            switch name {
            case "ble": value.insert(.bluetooth)
            case "timer": value.insert(.timer)
            case "persistence": value.insert(.persistence)
            case "secure_storage": value.insert(.secureStorage)
            case "network_transfer": value.insert(.networkTransfer)
            case "progress": value.insert(.progress)
            case "host_material": value.insert(.hostMaterial)
            case "recording_sink": value.insert(.recordingSink)
            case "firmware_blob": value.insert(.firmwareBlob)
            default:
                throw CoreError(
                    code: UInt32(BOTA_DEVICE_SDK_V1_ERROR_INVALID_INPUT),
                    operation: UInt32(BOTA_DEVICE_SDK_V1_OPERATION_DECODE),
                    retryable: false,
                    protocolStatus: nil,
                    detail: "unknown capability \(name)"
                )
            }
        }
        self = value
    }
}

struct CoreCancellationID: Equatable, Hashable, Sendable {
    let high: UInt64
    let low: UInt64

    init(_ id: UUID) {
        let bytes = withUnsafeBytes(of: id.uuid) { Array($0) }
        high = bytes[0..<8].reduce(0) { ($0 << 8) | UInt64($1) }
        low = bytes[8..<16].reduce(0) { ($0 << 8) | UInt64($1) }
    }
}

struct CoreCommand: Equatable, Sendable {
    let kind: UInt32
    let cancellationID: UUID
    let fields: [CoreField]

    var packet: CorePacket {
        let cancellation = CoreCancellationID(cancellationID)
        return CorePacket(
            kind: kind,
            operation: 0,
            requestID: 0,
            cancellationHigh: cancellation.high,
            cancellationLow: cancellation.low,
            fields: fields
        )
    }

    static func discoverDevices(
        timeoutMilliseconds: UInt64,
        allowDuplicates: Bool,
        cancellationID: UUID = UUID()
    ) -> Self {
        Self(
            kind: UInt32(BOTA_DEVICE_SDK_V1_COMMAND_DISCOVER_DEVICES),
            cancellationID: cancellationID,
            fields: [
                .unsigned(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_TIMEOUT_MS), value: timeoutMilliseconds),
                .bool(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_ALLOW_DUPLICATES), value: allowDuplicates),
            ]
        )
    }

    static func connect(
        serialNumber: String,
        peripheralID: String,
        name: String?,
        advertisedAddress: String?,
        rssi: Int16,
        cancellationID: UUID = UUID()
    ) -> Self {
        var fields: [CoreField] = [
            .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_SERIAL_NUMBER), value: serialNumber),
            .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_PERIPHERAL_ID), value: peripheralID),
            .signed(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_RSSI), value: Int64(rssi)),
        ]
        if let name { fields.append(.text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_NAME), value: name)) }
        if let advertisedAddress {
            fields.append(.text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_ADVERTISED_ADDRESS), value: advertisedAddress))
        }
        return Self(kind: UInt32(BOTA_DEVICE_SDK_V1_COMMAND_CONNECT), cancellationID: cancellationID, fields: fields)
    }

    static func connectSelected(
        peripheralID: String,
        name: String?,
        advertisedAddress: String?,
        rssi: Int16,
        cancellationID: UUID = UUID()
    ) -> Self {
        var fields: [CoreField] = [
            .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_PERIPHERAL_ID), value: peripheralID),
            .signed(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_RSSI), value: Int64(rssi)),
        ]
        if let name { fields.append(.text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_NAME), value: name)) }
        if let advertisedAddress {
            fields.append(.text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_ADVERTISED_ADDRESS), value: advertisedAddress))
        }
        return Self(kind: UInt32(BOTA_DEVICE_SDK_V1_COMMAND_CONNECT), cancellationID: cancellationID, fields: fields)
    }

    static func reconnect(
        serialNumber: String,
        storedPeripheralID: String? = nil,
        advertisedAddress: String? = nil,
        storedName: String? = nil,
        scanTimeoutMilliseconds: UInt64 = 10_000,
        connectionTimeoutMilliseconds: UInt64 = 10_000,
        cancellationID: UUID = UUID()
    ) -> Self {
        var fields: [CoreField] = [
            .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_SERIAL_NUMBER), value: serialNumber),
            .unsigned(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_SCAN_TIMEOUT_MS), value: scanTimeoutMilliseconds),
            .unsigned(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_CONNECTION_TIMEOUT_MS), value: connectionTimeoutMilliseconds),
        ]
        if let storedPeripheralID {
            fields.append(.text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_STORED_PERIPHERAL_ID), value: storedPeripheralID))
        }
        if let advertisedAddress {
            fields.append(.text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_ADVERTISED_ADDRESS), value: advertisedAddress))
        }
        if let storedName { fields.append(.text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_STORED_NAME), value: storedName)) }
        return Self(kind: UInt32(BOTA_DEVICE_SDK_V1_COMMAND_RECONNECT), cancellationID: cancellationID, fields: fields)
    }

    static func provision(serialNumber: String, materialID: String, cancellationID: UUID = UUID()) -> Self {
        Self(
            kind: UInt32(BOTA_DEVICE_SDK_V1_COMMAND_PROVISION),
            cancellationID: cancellationID,
            fields: [
                .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_SERIAL_NUMBER), value: serialNumber),
                .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_MATERIAL_ID), value: materialID),
            ]
        )
    }

    static func transferRecording(
        serialNumber: String,
        recordingUUID: String,
        sinkID: String,
        totalUnits: UInt64,
        confirmOnCompletion: Bool = true,
        cancellationID: UUID = UUID()
    ) -> Self {
        Self(
            kind: UInt32(BOTA_DEVICE_SDK_V1_COMMAND_TRANSFER_RECORDING),
            cancellationID: cancellationID,
            fields: [
                .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_SERIAL_NUMBER), value: serialNumber),
                .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_RECORDING_UUID), value: recordingUUID),
                .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_SINK_ID), value: sinkID),
                .unsigned(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_TOTAL_UNITS), value: totalUnits),
                .bool(id: 124, value: confirmOnCompletion),
            ]
        )
    }

    static func uploadRecording(
        serialNumber: String,
        recordingUUID: String,
        uploadID: String,
        destinationID: String,
        cancellationID: UUID = UUID()
    ) -> Self {
        Self(
            kind: UInt32(BOTA_DEVICE_SDK_V1_COMMAND_UPLOAD_RECORDING),
            cancellationID: cancellationID,
            fields: [
                .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_SERIAL_NUMBER), value: serialNumber),
                .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_RECORDING_UUID), value: recordingUUID),
                .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_UPLOAD_ID), value: uploadID),
                .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_DESTINATION_ID), value: destinationID),
            ]
        )
    }

    static func updateFirmware(
        serialNumber: String,
        version: String,
        sizeBytes: UInt32,
        crc32: UInt32,
        downloadID: UInt64,
        cancellationID: UUID = UUID()
    ) -> Self {
        Self(
            kind: UInt32(BOTA_DEVICE_SDK_V1_COMMAND_UPDATE_FIRMWARE),
            cancellationID: cancellationID,
            fields: [
                .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_SERIAL_NUMBER), value: serialNumber),
                .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_FIRMWARE_VERSION), value: version),
                .unsigned(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_FIRMWARE_SIZE_BYTES), value: UInt64(sizeBytes)),
                .unsigned(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_FIRMWARE_CRC32), value: UInt64(crc32)),
                .unsigned(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_DOWNLOAD_ID), value: downloadID),
            ]
        )
    }

    static func readDeviceLogs(serialNumber: String, cancellationID: UUID = UUID()) -> Self {
        Self(
            kind: UInt32(BOTA_DEVICE_SDK_V1_COMMAND_READ_DEVICE_LOGS),
            cancellationID: cancellationID,
            fields: [.text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_SERIAL_NUMBER), value: serialNumber)]
        )
    }

    static func factoryReset(
        serialNumber: String,
        commandID: String,
        grantID: String,
        cancellationID: UUID = UUID()
    ) -> Self {
        Self(
            kind: UInt32(BOTA_DEVICE_SDK_V1_COMMAND_FACTORY_RESET),
            cancellationID: cancellationID,
            fields: [
                .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_SERIAL_NUMBER), value: serialNumber),
                .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_COMMAND_ID), value: commandID),
                .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_GRANT_ID), value: grantID),
            ]
        )
    }

    static func resumeFactoryReset(
        serialNumber: String,
        commandID: String,
        resultCode: UInt8,
        deletedRecordingCount: UInt16,
        cancellationID: UUID = UUID()
    ) -> Self {
        Self(
            kind: UInt32(BOTA_DEVICE_SDK_V1_COMMAND_RESUME_FACTORY_RESET),
            cancellationID: cancellationID,
            fields: [
                .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_SERIAL_NUMBER), value: serialNumber),
                .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_COMMAND_ID), value: commandID),
                .unsigned(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_RESULT_CODE), value: UInt64(resultCode)),
                .unsigned(
                    id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_DELETED_RECORDING_COUNT),
                    value: UInt64(deletedRecordingCount)
                ),
            ]
        )
    }

    static func resumeUnjournaledFactoryReset(
        serialNumber: String,
        commandID: String,
        cancellationID: UUID = UUID()
    ) -> Self {
        Self(
            kind: UInt32(BOTA_DEVICE_SDK_V1_COMMAND_RESUME_FACTORY_RESET),
            cancellationID: cancellationID,
            fields: [
                .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_SERIAL_NUMBER), value: serialNumber),
                .text(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_COMMAND_ID), value: commandID),
            ]
        )
    }
}

extension CoreCommand {
    static func fixture(named name: String) -> Self? {
        let serial = "EVFXXW67KP"
        let recording = "00112233445566778899aabbccddeeff"
        switch name {
        case "discover_devices": return .discoverDevices(timeoutMilliseconds: 1, allowDuplicates: false)
        case "connect": return .connect(serialNumber: serial, peripheralID: "peripheral", name: nil, advertisedAddress: nil, rssi: -40)
        case "reconnect": return .reconnect(serialNumber: serial)
        case "provision": return .provision(serialNumber: serial, materialID: "material")
        case "transfer_recording": return .transferRecording(serialNumber: serial, recordingUUID: recording, sinkID: "sink", totalUnits: 1)
        case "upload_recording": return .uploadRecording(serialNumber: serial, recordingUUID: recording, uploadID: "upload", destinationID: "destination")
        case "update_firmware": return .updateFirmware(serialNumber: serial, version: "1.0.0", sizeBytes: 1, crc32: 1, downloadID: 1)
        case "read_device_logs": return .readDeviceLogs(serialNumber: serial)
        case "factory_reset": return .factoryReset(serialNumber: serial, commandID: "command", grantID: "grant")
        case "resume_factory_reset": return .resumeFactoryReset(serialNumber: serial, commandID: "command", resultCode: 0, deletedRecordingCount: 0)
        default: return nil
        }
    }
}

File: platforms/android/sdk/src/main/kotlin/dev/bota/sdk/internal/core/CoreCommand.kt
package dev.bota.sdk.internal.core

import dev.bota.sdk.internal.jni.NativePacket
import java.util.UUID

internal sealed interface CoreField {
    val id: Int

    data class Unsigned(override val id: Int, val value: ULong) : CoreField
    data class Signed(override val id: Int, val value: Long) : CoreField
    data class BooleanValue(override val id: Int, val value: Boolean) : CoreField
    data class Text(override val id: Int, val value: String) : CoreField
    class Bytes(override val id: Int, value: ByteArray) : CoreField {
        val value: ByteArray = value.copyOf()

        override fun equals(other: Any?): Boolean =
            other is Bytes && id == other.id && value.contentEquals(other.value)

        override fun hashCode(): Int = 31 * id + value.contentHashCode()
    }
}

internal data class CoreCapabilities(val bits: ULong) {
    operator fun plus(other: CoreCapabilities): CoreCapabilities = CoreCapabilities(bits or other.bits)

    companion object {
        val Bluetooth = CoreCapabilities(1uL shl 0)
        val Timer = CoreCapabilities(1uL shl 1)
        val Persistence = CoreCapabilities(1uL shl 2)
        val SecureStorage = CoreCapabilities(1uL shl 3)
        val NetworkTransfer = CoreCapabilities(1uL shl 4)
        val Progress = CoreCapabilities(1uL shl 5)
        val HostMaterial = CoreCapabilities(1uL shl 6)
        val RecordingSink = CoreCapabilities(1uL shl 7)
        val FirmwareBlob = CoreCapabilities(1uL shl 8)

        fun fromNames(names: List<String>): CoreCapabilities = names.fold(CoreCapabilities(0u)) { value, name ->
            value + when (name) {
                "ble" -> Bluetooth
                "timer" -> Timer
                "persistence" -> Persistence
                "secure_storage" -> SecureStorage
                "network_transfer" -> NetworkTransfer
                "progress" -> Progress
                "host_material" -> HostMaterial
                "recording_sink" -> RecordingSink
                "firmware_blob" -> FirmwareBlob
                else -> throw IllegalArgumentException("unknown capability $name")
            }
        }
    }
}

internal data class CoreCancellationId(val high: ULong, val low: ULong) {
    constructor(id: UUID) : this(id.mostSignificantBits.toULong(), id.leastSignificantBits.toULong())
}

internal data class CoreCommand(
    val kind: Int,
    val cancellationId: UUID,
    val fields: List<CoreField>,
) {
    val packet: NativePacket
        get() = fields.toNativePacket(kind, cancellationId = CoreCancellationId(cancellationId))

    companion object {
        fun discoverDevices(
            timeoutMilliseconds: ULong,
            allowDuplicates: Boolean,
            cancellationId: UUID = UUID.randomUUID(),
        ) = CoreCommand(
            0x0101,
            cancellationId,
            listOf(CoreField.Unsigned(1, timeoutMilliseconds), CoreField.BooleanValue(2, allowDuplicates)),
        )

        fun connect(
            serialNumber: String,
            peripheralId: String,
            name: String?,
            advertisedAddress: String?,
            rssi: Int,
            cancellationId: UUID = UUID.randomUUID(),
        ) = CoreCommand(
            0x0102,
            cancellationId,
            buildList {
                add(CoreField.Text(3, serialNumber))
                add(CoreField.Text(4, peripheralId))
                add(CoreField.Signed(7, rssi.toLong()))
                name?.let { add(CoreField.Text(5, it)) }
                advertisedAddress?.let { add(CoreField.Text(6, it)) }
            },
        )

        fun connectSelected(
            peripheralId: String,
            name: String?,
            advertisedAddress: String?,
            rssi: Int,
            cancellationId: UUID = UUID.randomUUID(),
        ) = CoreCommand(
            0x0102,
            cancellationId,
            buildList {
                add(CoreField.Text(4, peripheralId))
                add(CoreField.Signed(7, rssi.toLong()))
                name?.let { add(CoreField.Text(5, it)) }
                advertisedAddress?.let { add(CoreField.Text(6, it)) }
            },
        )

        fun reconnect(
            serialNumber: String,
            storedPeripheralId: String? = null,
            advertisedAddress: String? = null,
            storedName: String? = null,
            scanTimeoutMilliseconds: ULong = 10_000u,
            connectionTimeoutMilliseconds: ULong = 10_000u,
            cancellationId: UUID = UUID.randomUUID(),
        ) = CoreCommand(
            0x0103,
            cancellationId,
            buildList {
                add(CoreField.Text(3, serialNumber))
                add(CoreField.Unsigned(10, scanTimeoutMilliseconds))
                add(CoreField.Unsigned(11, connectionTimeoutMilliseconds))
                storedPeripheralId?.let { add(CoreField.Text(8, it)) }
                advertisedAddress?.let { add(CoreField.Text(6, it)) }
                storedName?.let { add(CoreField.Text(9, it)) }
            },
        )

        fun provision(serialNumber: String, materialId: String, cancellationId: UUID = UUID.randomUUID()) =
            command(0x0104, cancellationId, CoreField.Text(3, serialNumber), CoreField.Text(12, materialId))

        fun transferRecording(
            serialNumber: String,
            recordingUuid: String,
            sinkId: String,
            totalUnits: ULong,
            confirmOnCompletion: Boolean = true,
            cancellationId: UUID = UUID.randomUUID(),
        ) = command(
            0x0105,
            cancellationId,
            CoreField.Text(3, serialNumber),
            CoreField.Text(13, recordingUuid),
            CoreField.Text(14, sinkId),
            CoreField.Unsigned(15, totalUnits),
            CoreField.BooleanValue(124, confirmOnCompletion),
        )

        fun uploadRecording(
            serialNumber: String,
            recordingUuid: String,
            uploadId: String,
            destinationId: String,
            cancellationId: UUID = UUID.randomUUID(),
        ) = command(
            0x0106,
            cancellationId,
            CoreField.Text(3, serialNumber),
            CoreField.Text(13, recordingUuid),
            CoreField.Text(16, uploadId),
            CoreField.Text(17, destinationId),
        )

        fun updateFirmware(
            serialNumber: String,
            version: String,
            sizeBytes: UInt,
            crc32: UInt,
            downloadId: ULong,
            cancellationId: UUID = UUID.randomUUID(),
        ) = command(
            0x0107,
            cancellationId,
            CoreField.Text(3, serialNumber),
            CoreField.Text(18, version),
            CoreField.Unsigned(19, sizeBytes.toULong()),
            CoreField.Unsigned(20, crc32.toULong()),
            CoreField.Unsigned(21, downloadId),
        )

        fun readDeviceLogs(serialNumber: String, cancellationId: UUID = UUID.randomUUID()) =
            command(0x0108, cancellationId, CoreField.Text(3, serialNumber))

        fun factoryReset(
            serialNumber: String,
            commandId: String,
            grantId: String,
            cancellationId: UUID = UUID.randomUUID(),
        ) = command(
            0x0109,
            cancellationId,
            CoreField.Text(3, serialNumber),
            CoreField.Text(22, commandId),
            CoreField.Text(23, grantId),
        )

        fun resumeFactoryReset(
            serialNumber: String,
            commandId: String,
            resultCode: UByte,
            deletedRecordingCount: UShort,
            cancellationId: UUID = UUID.randomUUID(),
        ) = command(
            0x010a,
            cancellationId,
            CoreField.Text(3, serialNumber),
            CoreField.Text(22, commandId),
            CoreField.Unsigned(24, resultCode.toULong()),
            CoreField.Unsigned(25, deletedRecordingCount.toULong()),
        )

        fun resumeUnjournaledFactoryReset(
            serialNumber: String,
            commandId: String,
            cancellationId: UUID = UUID.randomUUID(),
        ) = command(
            0x010a,
            cancellationId,
            CoreField.Text(3, serialNumber),
            CoreField.Text(22, commandId),
        )

        fun fixtureNamed(name: String): CoreCommand? {
            val serial = "EVFXXW67KP"
            val recording = "00112233445566778899aabbccddeeff"
            return when (name) {
                "discover_devices" -> discoverDevices(1u, false)
                "connect" -> connect(serial, "peripheral", null, null, -40)
                "reconnect" -> reconnect(serial)
                "provision" -> provision(serial, "material")
                "transfer_recording" -> transferRecording(serial, recording, "sink", 1u)
                "upload_recording" -> uploadRecording(serial, recording, "upload", "destination")
                "update_firmware" -> updateFirmware(serial, "1.0.0", 1u, 1u, 1u)
                "read_device_logs" -> readDeviceLogs(serial)
                "factory_reset" -> factoryReset(serial, "command", "grant")
                "resume_factory_reset" -> resumeFactoryReset(serial, "command", 0u, 0u)
                else -> null
            }
        }

        private fun command(kind: Int, cancellationId: UUID, vararg fields: CoreField) =
            CoreCommand(kind, cancellationId, fields.toList())
    }
}

internal fun List<CoreField>.toNativePacket(
    kind: Int,
    operation: Int = 0,
    requestId: ULong = 0u,
    cancellationId: CoreCancellationId = CoreCancellationId(0u, 0u),
): NativePacket {
    val types = IntArray(size)
    val unsigned = LongArray(size)
    val signed = LongArray(size)
    val data = arrayOfNulls<Any>(size)
    forEachIndexed { index, field ->
        when (field) {
            is CoreField.Unsigned -> {
                types[index] = NativePacket.FIELD_TYPE_UNSIGNED
                unsigned[index] = field.value.toLong()
            }
            is CoreField.Signed -> {
                types[index] = NativePacket.FIELD_TYPE_SIGNED
                signed[index] = field.value
            }
            is CoreField.BooleanValue -> {
                types[index] = NativePacket.FIELD_TYPE_BOOL
                unsigned[index] = if (field.value) 1 else 0
            }
            is CoreField.Text -> {
                types[index] = NativePacket.FIELD_TYPE_UTF8
                data[index] = field.value.encodeToByteArray()
            }
            is CoreField.Bytes -> {
                types[index] = NativePacket.FIELD_TYPE_BYTES
                data[index] = field.value.copyOf()
            }
        }
    }
    return NativePacket(
        kind = kind,
        operation = operation,
        requestIdBits = requestId.toLong(),
        cancellationHighBits = cancellationId.high.toLong(),
        cancellationLowBits = cancellationId.low.toLong(),
        fieldIds = map(CoreField::id).toIntArray(),
        fieldTypes = types,
        unsignedValues = unsigned,
        signedValues = signed,
        dataValues = data,
    )
}
